src/query/service/src/auth.rs

```diff
@@ -78,34 +80,47 @@ impl AuthMgr {
                     session.set_current_tenant(tenant);
                 };
 
-                // create a virtual JWT user only available in current session
-                let auth_role = jwt.custom.role.clone();
-                let mut user_info = UserInfo::new(&user_name, "%", AuthInfo::JWT);
-                if user_info.identity().is_root() {
-                    return Err(ErrorCode::AuthenticateFailure(
-                        "root user is not allowed in jwt auth.",
-                    ));
-                }
-                if let Some(ref role) = auth_role {
-                    user_info.grants.grant_role(role.clone());
-                }
-                if let Some(ref ensure_user) = jwt.custom.ensure_user {
-                    if let Some(ref roles) = ensure_user.roles {
-                        for role in roles.clone().into_iter() {
-                            user_info.grants.grant_role(role);
+                let tenant = session.get_current_tenant();
+                let identity = UserIdentity::new(&user_name, "%");
+                let user = match user_api.get_user(&tenant, identity.clone()).await {
+                    Ok(user_info) => match user_info.auth_info {
+                        AuthInfo::JWT => Ok(user_info),
+                        _ => return Err(ErrorCode::AuthenticateFailure("wrong auth type")),
+                    },
+                    Err(e) => {
+                        if e.code() != ErrorCode::UNKNOWN_USER {
+                            return Err(ErrorCode::AuthenticateFailure(e.message()));
+                        }
+                        let ensure_user = jwt
+                            .custom
+                            .ensure_user
+                            .ok_or(ErrorCode::AuthenticateFailure(e.message()))?;
+                        // create a new user if not exists
+                        let mut user_info = UserInfo::new(&user_name, "%", AuthInfo::JWT);
+                        if let Some(ref roles) = ensure_user.roles {
+                            for role in roles.clone().into_iter() {
+                                user_info.grants.grant_role(role);
+                            }
                         }
+                        user_api.add_user(&tenant, user_info, true).await?;
+                        user_api.get_user(&tenant, identity).await
                     }
-                }
+                }?;
+
+                let user = match &user.auth_info {
+                    AuthInfo::JWT => user,
+                    _ => return Err(ErrorCode::AuthenticateFailure("wrong auth type")),
+                };
```

Comment:
I guess this check is duplicated with L86: https://github.com/datafuselabs/databend/pull/10786/files#diff-fe9c4d4d86fb6815af28d24728a7e98cc84952c53d2edd270a76bdd7279cd333R86